### ISSUE TITLE
Random number generator library

### ISSUE THREAD

**Issue by maximecb** [zetavm/zetavm#39]
It would be useful to have a random number generator library. Something we could use as follows:

```
var random = import 'std/random';
var rng = random.newRNG(seedVal);

// Generate an integer in [a,b]
var intVal = rng.int(a,b);

// Generate a float in [a,b]
var floatVal = rng.float(a,b);

// Generate a random index in [0, len[
var idx = rng.index(len);
```

Would also be useful to have a method to select a random element in an array, and to generate random booleans, etc.

I've written this C++ code a while back which could be ported, serve as a base for the implementation: https://gist.github.com/maximecb/617de45a99347a9911b1e0d974da5d62

**Comment by tommyettinger** [comment]
This is something to be... careful about. A good cautionary reference is .NET's System.Random implementation, which is known to have improperly implemented the algorithm it states it uses in its documentation, but must stay using that incorrect, possibly-significantly-flawed algorithm because users are relying on given seeds producing the same output between .NET versions. References for the bug and its won't-fix status here: https://connect.microsoft.com/VisualStudio/feedback/details/634761/system-random-serious-bug . I would encourage selecting a generator from http://xoroshiro.di.unimi.it/ that fits the features you want in a PRNG. [PCG-Random](http://www.pcg-random.org/) may not seem like the best choice from the speed/failures graph there but it has a lot of other optional features that may sway in its favor, including various state sizes and configurable extra features, some of which entail ludicrously good quality when the state is very large. Similarly, xoroshiro128+ (linked on the unimi.it page) usually has great speed, but mainly if the compiler can produce SSE (or similar processor extension) bitwise rotation instructions, and probably will not do as well if that can't be relied on. SplitMix64 (linked on the unimi.it page) is small enough that it can be written inline as part of the initialization of another generator, which is a good idea, plus it has some other useful traits. If 64-bit math can't be relied on to be fast, then the playing field completely changes, and PCG-Random's various setups that use 32-bit values for state will probably blitz past xoroshiro128+ and SplitMix64. It should be mentioned that xoroshiro128+, like a linear congruential generator, does not tolerate a state of 0 for all state values, while PCG-Random does tolerate all states for at least some setups, as does SplitMix64 (this isn't a common feature, but can be very nice when randomly seeding a generator to know whatever numbers are given, it won't produce an invalid state). Other features to keep in mind are skip-ahead/skip-back functionality, which PCG can do in less than linear time (for some setups, it should be constant) and SplitMix64 can do in constant time. PCG-Random has implementations in C and C++, some of which are very concise, while SplitMix64 and xoroshiro128+ are each one small C file, or a part of one.

I implement various PRNGs, some of them novel, for fun in Java and C#, and while some of them have better speed than the fastest generators tested at http://xoroshiro.di.unimi.it/ (comparing JVM implementations to JVM implementations), I know my toy generators have worse quality and uncertain periodicity. I'd be very reluctant to use one of my toy generators for any critical purpose, though I'm fine using them in games. I've found that platform quirks have a tremendous effect on the relative performance of these PRNGs. SplitMix64 is tied with xoroshiro128+ for speed on a 64-bit desktop version of the JVM, which is unexpected considering their performance is very different in C. In C#, SplitMix64 performs well on 64-bit builds (better than twice the speed of System.Random, with a better period, and without the quality/period issues present in most linear congruential generators), but abysmally on 32-bit, measurably worse than System.Random (which uses 32-bit math only).

I'm looking forward to watching ZetaVM develop! I'm curious about the qualities that may emerge in some new form in a VM built for dynamic languages.

**Comment by maximecb** [comment]
The one upside we have is that every library in ZetaVM is going to be versioned. So we can keep version zero of the random number generator, even though we supersede it with a patched version.

For the default RNG, I want something that will be reasonably fast. It doesn't need to be crypto-safe.  It needs to be fast because there are applications such as raytracing where this is important.

Would you be interested in writing a simple PRNG for Zeta? :)

**Comment by tommyettinger** [comment]
Sure thing. I haven't even cloned the project yet, and I don't have any idea what this would entail -- would this PRNG make sense to be written in Zeta's textual bytecode? In C/C++? In Plush and compiled to a .zim file? A possible issue might be that I primarily develop on Windows (7), and I've hit issues before with VMs and other language-development projects (most recently, a project that seems to be funded by Oracle, [Graal](https://github.com/graalvm/graal) ) where support for the OS variety installed on > 90% of desktops and laptops is non-existent. The approach that I think Zeta is taking, where the textual bytecode can be produced by anything and consumed anywhere Zeta runs, looks like one of the few good ways to handle complicated multi-language projects and have them all able to communicate through simple files. The analogy to JSON seems apt, since it seems to be used for a similar purpose frequently.

But I'm rambling. I'll try to build Zeta now.

**Comment by maximecb** [comment]
The easiest way would be to code it in the Plush language (.pls files). It's kind of like JavaScript-lite. You could write the bytecode directly, but it might be kind of a pain. I think that eventually we might be able to compile C/C++ or even machine code into the Zeta IR, so that we can support existing code, but that will take a while.

Right now everything is interpreter, there is no JIT compiler, so it should be possible to run on windows, provided you can run make, gcc, etc.

**Comment by tommyettinger** [comment]
It's mostly on its way to working on Windows. Most of the file reading/opening operations are sensitive to the CRLF line endings that are hard to avoid, so I appended a `b` mode modifier to all the `fopen` operations in my fork, https://github.com/tommyettinger/zetavm , which makes it use LF on Cygwin at least. It's still not passing `make test`; it fails trying to call `math.sin` with the error `aborting execution due to error: unhandled base type in read of property "sin"` in `tests/plush/floats.pls`, it looks like on line 33 or 34.

**Comment by maximecb** [comment]
Hm. I don't have a windows machine to test this on, but I thought the code would have handled it.

**Comment by Maaarcocr** [comment]
@maximecb we should add AppVeyor (it's similar to travis) to build and test on Windows. I will have a look on how to do that!

**Comment by tommyettinger** [comment]
Actually it isn't the text editor that causes the problem, surprisingly. zetavm didn't have a `.gitattributes` file in the version I forked, so it would use the platform defaults, which are usually "use CRLF when checking out on Windows, commit as LF". Unfortunately, the C and/or C++ standard library functions for dealing with files seem not to like CRLF when Cygwin is in use and binary mode (Unix's default) isn't specified by methods like `fopen()`, and give different counts for lines with `ftell()` and the result of `fread()`. This also affects the behavior of shell redirection using `>` , so without it redirecting to a new file would simply produce that file in text mode (so, using CRLF on Windows). I'll submit a PR with the relatively minor changes I needed to make, mostly to configuration but sometimes to code.

**Comment by tommyettinger** [comment]
I'm making progress. I can update my Zeta/Plush code now and test it on this Windows machine, and the next step is getting all the bitwise mechanics in-place. The PR I just sent, referenced above, should add most of the bitwise operators that I can think of. There are some potential issues in corner cases, but generally this would be a problem only if people start pushing the limits of normal bitwise operations. Since char values in Zeta seem to be, at least some of the time, 32-bit signed integers that only use the bottom 16 bits, using the `~` bitwise one's-complement operator would produce a negative number because the sign bit would be flipped, though the bottom 16 bits would remain correctly inverted. My favorite example of this char-bit-twiddling is the painfully-complex, but valid, Java code `char[] $$$ = {'$' << (',' ^ '-'), '`' | ('-' - '('), '$' << (',' ^ '-') | '$', '$' << (',' ^ '-') | '$', '/' + '@', ',', ' ', '(' + '/', '/' + '@', '(' + '%' + '%', '$' << (',' ^ '-') | '$', '*' + ':', '!'};`, which produces a char array called `$$$` containing "Hello, World!" without using any alphanumeric chars other than those in keywords, and achieves this mainly through bitwise operations on char literals. I've occasionally used this novelty for actual benefit, mainly in high-speed cross-platform hashing functions that need to operate on char arrays as well as other arrays. I think I may have also used it in an int-to-string or string-to-int function that fit certain less-commonly-implemented requirements. I believe the unsigned right shift operator, `>>>` is implemented correctly, and it passes at least one simple test, but I'm not 100% sure it will work as intended. The sign-extending right shift operator, `>>`. similarly seems to work... but I have heard anecdotally that right-shift on signed ints is undefined in some versions of C or C++.  The bitwise and (`&`), or (`|`), and xor (`^`) operators all appear to work on either pairs of ints or pairs of booleans, which seems like the correct behavior (while you could mimic `boolean ^ boolean` with `boolean != boolean`, that way couldn't be mixed with an assignment, contrasted with `^=`).

I should ask, *are assignments usable as expressions in Plush?* I'm not sure if this behavior is considered desirable or not; C, C++, Java, and C# all have it but the last two use it very rarely, while Kotlin copies most of Java's design decisions except for intentionally making assignments unusable as expressions. The reason I ask is that the central part of the random number generator I have uses this somewhat extensively, and a subtle quirk of this behavior may contribute to a higher-than-expected quality. The entire body of the method that gets the next int looks like `(state[(choice += 0xCBBC475B) & 31] += (state[choice >>> 28] += 0x9C7B7B99) >>> 1)` , which may look like a cat walked on my keyboard but actually needed quite a lot of fine-tuning. In that code, choice is an int and state is a 32-item int array. The strange part of this, and the one where assignments matter as expressions, is that this chooses  to update one part of state selected using the bottom 5 bits of choice, and updates another part of state selected using the upper 4 bits of choice. These may sometimes be the same element of state, in which case the amount changed is not what it might appear at first glance. If we simplified the code so that the constant is the example value 111, the relevant upper and lower sections of choice are each equal to 2 in this iteration, then the code looks like `(state[2] += (state[2] += 111) >>> 1)`. The quirk here is that parentheses don't act quite as you'd expect in grouping precedence with assignments, at least in Java and maybe also in C/C++, and if `state[2]` started with 1000 as its value, then it wouldn't be `(state[2] += (state[2] += 111) >>> 1)` (state2 is 1000) => `(state[2] += 1111 >>> 1)` (state2 is 1111) => `(state[2] += 555)` (state2 is still 1111) => 1666 (state2 is 1666), but would instead be `(state[2] += (state[2] += 111) >>> 1)` (state2 is 1000) => `(state[2] += 1111 >>> 1)` (state2 is 1111 for now, but the value it had when the `+=` assignment started, 1000, will still be used for that assignment) => `(state[2] += 555)` (now there's nothing but the value the assignment started with, so state2 will use 1000 for its value) => 1555 (state2 is 1555). This may be subtly incorrect information, and I've had a hard time finding specific answers to this type of question in Java. It might actually be easier to get an understanding of how I could replicate this behavior in Zeta's clear source than to find a correct answer somewhere on the Internet that explains how Java does it...

**Comment by maximecb** [comment]
Happy about the bitwise ops PR.

I believe assignments are usable as expressions, but I personally would avoid writing code that does this, as it makes the code harder to read.

**Comment by tommyettinger** [comment]
Almost there... I submitted PR #49 to finish the existing code that sets elements in arrays. Now I'm not limited to a PRNG with a period of 2 to the 32, and can use an array of state values to improve the period and quality.

Should this go in a file called `plush/random.pls` ? Is there a different place it should go?

**Comment by maximecb** [comment]
@tommyettinger Awesome!

You can put it in packages/std/random/0/package. This will be importable in code with `var random = import "std/random/0"`.

**Comment by maximecb** [comment]
@tommyettinger Curious to know how you think your RNG will perform. Is the size of the state tunable? I'm asking because I might need to generate noise on the fly for audio generation.

I guess we can add an RNG benchmark to our set when you have it working :)

**Comment by tommyettinger** [comment]
Tunable state size would be nice. I'm currently debating either:
  - Defaulting to a 1024-bit state array, or 32 int32 in an array plus another outside of it. I have found good constant values to use, but they're just good, not great (with them, it passes many quality tests, but not all, with two failures on PractRand's 32-bit folding mode).
  - Defaulting to a larger state array if I can find a technique that can pass all the tests, but allowing a smaller or larger state size through some configuration, possibly selecting from the best candidates known at specific sizes.

Either choice will also allow a separate way to get a pseudo-random int32 from an int32 parameter that is expected to be changed with each call; this is used if it needs to seed the state given less (or more) bits of input than the state size. This probably will also be wrapped up as its own type of random generator, usable on its own for applications that don't need a lot of period (This has a period of 2 to the 32 if an odd-number increment is added to the parameter each time). It's possible I can find a good way to allow more-arbitrary state sizes than power-of-two multiples of 32 bits, but that would likely do best with larger states to ensure reliable quality.

**Comment by tommyettinger** [comment]
OK, PR submitted. The version that I wound up using in the PR has about a 2080-bit state (could also be considered 2048-bit), and passes PractRand across the board with nothing worse than 1 or 2 anomalies out of over a hundred tests, and they're classified as merely "unusual", rather than "suspicious", "very suspicious", "VERY SUSPICIOUS", or "FAIL".

It isn't the fastest algorithm, but that isn't exactly an issue when you compare it to what some other standard libraries use (BSD's system `rand()` function uses a full-blown cryptographic cipher applied to a counter, while testing this 2080-bit algorithm in Java has this generator about 5x to 15x faster than Java's built-in `java.util.Random`, and sometimes on-par with the newer `java.util.SplittableRandom`). The only algorithms I have implemented that are much faster also have severe quality problems and lower periods, so I'm pleased with this result. JIT compilation may have significant effects on my tests in Java, since it goes from 2 seconds to generate the first billion 32-bit values to 1.5 seconds to generate any future chunks of a billion numbers, and testing with one layer of indirection slows it just below 2 seconds for all runs (likely due to Java's JIT compilation being unable to optimize in some way that it can when less indirection is being used). To contrast, `java.util.Random`, which is a simple LCG on 64-bit numbers that drops the bottom 22 or 16 (not sure) bits for quality, takes quite a bit longer; on my machine I think it took a minimum of 10 seconds, likely longer and probably about 15-20 seconds, to generate a billion 32-bit values. A normal LCG wouldn't take that long; Java's `Random` class has a lot of relative overhead by being synchronized for use on multiple threads. I'd suspect this package will be a little slower than a 32-bit-state LCG (optimization can probably take it to better than 50% of the speed of an LCG if it isn't there yet), but won't have low period on the less-significant bits, will have a (much) higher period than the 2 to the 32 minus 1 that an LCG would be limited to with 32-bit math as the only option, and also passes statistical tests that LCGs with low amounts of state struggle with.

**Comment by maximecb** [comment]
I'm happy to see the end result! It's nice that you took the time to rigorously test the algorithm as well. Curious to know how you came to acquire your RNG expertise :)

**Comment by tommyettinger** [comment]
It's a good question... I started off wanting to improve some speed problems an existing library was having with a bad, extremely-slow multi-threaded Mersenne Twister implementation, then implementing ,many of the better generators at [this PRNG benchmark site](http://xoroshiro.di.unimi.it/) in Java for the library, then moving on to hashing algorithms, getting inspiration there and moving back to RNGs, and so on... I read Melissa E. O'Neill's paper on PCG-Random so I could implement her nice algorithms, read only a few others (L'Ecuyer has some important works that I glanced through, then later I read a bit on non-linear feedback shift registers. NLFSR work is brand-new and meant for cryptographic random usage, but could have other uses.). I discovered that data dependency can slow down some code by a large degree, and so paradoxically having two states that update independently can be more efficient than having one state that needs its earlier value during a calculation to assign it a new value. At least in Java, `x += y * 9` seems to sidestep the data dependency issue by only using x's previous value at the very end of the calculation, while `x += x * 9` can't do the same. This is all handled at the processor level and depends on how the processor is able to optimize the code. Knowing about data dependency had me writing new RNGs to exploit this optimization, where the new RNGs had terrible quality and period but great speed, up to twice as fast as an LCG. At some point I decided to find out just how bad these new ones were on quality and started using PractRand to test quality, because the main tests for RNGs, DIEHARD and DIEHARDER, are hard to add new tests to and work best on Linux. The one I am using here is the second I've made that seems to never fail statistical tests outright with PractRand; the other uses an array of 64 64-bit values instead of an array of 64 32-bit values, but has very different constants and slightly different behavior.

Ramble... off.
